### DISCUSSION TITLE
How to use JPH::Mutex

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1050] [Q&A] [answered]
Sorry for the poor title, I'm not sure how to properly word what I'm getting at here...

During the contact callbacks, I need to cache these in a list for single-threaded processing later... in the contact callback, I only do the bare minimum (i.e., set the restitution, etc), trying to keep things thread-safe.

Would this be a proper use of the JPH::Mutex to ensure that adding to my cache list occurs atomically? If not, is there a way you recommend to achieve this?

Listable::AddFront() simply adds an item to the front of a doubly-linked list (&interactions points to the first item in the list).

I'm not married to using my Listable() class, either. If there's an atomic C++ collection that I can use instead, that is also sufficient.

`

JPH::Mutex CollisionHandler::interactionMutex;
CollisionInteraction *CollisionHandler::interactions = NULL;
void CollisionHandler::AddInteraction(CollisionInteraction *interaction)
{
	JPH::lock_guard lock(interactionMutex); // Does this keep this scope from being executed more than once at a time on all threads?
	Listable::AddFront(interaction, (Listable **)&interactions);
}

`

**Comment by jrouwe** [comment, marked as answer]
Your suggested code should work and I would do it that way until a profiler tells you that you spend a lot of time waiting for `interactionMutex`. If this is the case you can always build a lock free linked list (but I don't think it will be needed).

**Reply by SSNTails** [reply]
Thanks -- it's been working so far without any race conditions on Win32, but I was curious if issues would start cropping up on another platform (or even just another computer). 👍 Glad to see I'm using it the right way.
